File: LLWallPaper.App/Services/RotationService.cs
using LLWallPaper.App.Models;
using LLWallPaper.App.Utils;

namespace LLWallPaper.App.Services;

public sealed class RotationService
{
    private readonly Random _random = new();

    public CardItem? PickNext(
        IReadOnlyList<CardItem> candidates,
        IReadOnlyCollection<string> recentKeys,
        IReadOnlyCollection<string> favoriteKeys,
        IReadOnlyCollection<string> blockedKeys,
        bool preferFavorites,
        bool excludeBlocked,
        bool excludeThirdEvolution,
        bool excludeSrCards
    )
    {
        IEnumerable<CardItem> pool = candidates;
        if (excludeBlocked && blockedKeys.Count > 0)
        {
            pool = pool.Where(card => !blockedKeys.Contains(card.Id));
        }

        if (excludeThirdEvolution)
        {
            pool = pool.Where(card => !card.Id.EndsWith("2", StringComparison.Ordinal));
        }

        if (excludeSrCards)
        {
            pool = pool.Where(card => !CharacterMap.IsSrCard(card.Id));
        }

        if (recentKeys.Count > 0)
        {
            pool = pool.Where(card => !recentKeys.Contains(card.Id));
        }

        var filtered = pool.ToList();
        if (filtered.Count == 0)
        {
            return null;
        }

        if (preferFavorites)
        {
            var favorites = filtered.Where(card => favoriteKeys.Contains(card.Id)).ToList();
            if (favorites.Count > 0)
            {
                return favorites[_random.Next(favorites.Count)];
            }
        }

        return filtered[_random.Next(filtered.Count)];
    }
}

File: LLWallPaper.Mac/Sources/LLWallPaperMacCore/Domain/RotationService.swift
import Foundation

public struct RotationService: Sendable {
  public init() {}

  public func pickNext(
    candidates: [CardItem],
    recentKeys: Set<String>,
    favoriteKeys: Set<String>,
    blockedKeys: Set<String>,
    settings: AppSettings
  ) -> CardItem? {
    var pool = candidates

    if settings.excludeBlocked {
      pool = pool.filter { !blockedKeys.contains($0.id) }
    }

    if settings.excludeThirdEvolution {
      pool = pool.filter { !$0.id.hasSuffix("2") }
    }

    if settings.excludeSrCards {
      pool = pool.filter { !CharacterMap.isSrCard($0.id) }
    }

    if !recentKeys.isEmpty {
      pool = pool.filter { !recentKeys.contains($0.id) }
    }

    if pool.isEmpty {
      return nil
    }

    if settings.preferFavorites {
      let favorites = pool.filter { favoriteKeys.contains($0.id) }
      if let favorite = favorites.randomElement() {
        return favorite
      }
    }

    return pool.randomElement()
  }
}
